### DISCUSSION TITLE
Control Motor using radio and Pico W

### DISCUSSION THREAD

**Discussion by GenghisHan-op** [pybricks/support#1938] [General]
Hi,
I tried to control a motor ( attached to Port C of a Technic Hub) via 27Mhz radio control.
The radio receiver board is connected to my Pico, and the Pico will read the signal from it and broadcast to the Hub by BLE to control motor. The pico is placed next to the Hub.
The pi broadcast the signal with no problem but the hub is very unstable when receiving.
Here is my code on the pi
```from machine import Pin
import time
from bleradio import BLERadio

Fpin = machine.Pin(16, machine.Pin.IN)
Bpin = machine.Pin(17, machine.Pin.IN)

radio = BLERadio(broadcast_channel=5, observe_channels=[4, 18])
while True:
    if Fpin.value() == 1:
        radio.broadcast([1,0])
    else:
        radio.broadcast("0")
    if Bpin.value() == 1:
        radio.broadcast([0,1])
    else:
        radio.broadcast("0")
        time.sleep(0.1)
        
```
and here is the code from the hub
```from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Color, Port
from pybricks.tools import wait

hub = TechnicHub(observe_channels=[5])
motor = Motor(Port.C)
while True:
    mainstatus = 0
    data = hub.ble.observe(5)
    if data is None:
        hub.light.on(Color.RED)
    else:
        hub.light.on(Color.GREEN)
        print(data)
        if data == (1,0):
            motor.dc(100)  
        if data == (0,1):
            motor.dc(-100)
        if data == '0':
            motor.dc(0)
    wait(100)
```
When I press and hold the button, here is the received data:
```(0, 1)
(0, 1)
0
0
0
(0, 1)
(0, 1)
(0, 1)
0
0
0
(0, 1)
(0, 1)
```
It's very unstable.
Can anyone have any solution ?

**Comment by laurensvalk** [comment]
Have a look at your main sending code.

It is always broadcasting 0 even if you hold one of the buttons. So your are broadcasting `(0, 1)` and `0` in a very tight loop. The output given is to be expected.

You might want to change the sending logic to only change the broadcast signal when the button state changes, rather than re-broadcasting the same values over and over.

For example, you could:
- send 0
- wait until any button pressed
- send [1, 0] if button F pressed, else send [0, 1]
- wait until no buttons pressed

**Reply by GenghisHan-op** [reply]
Thanks for your reply,
Now I can see the problem in my broadcast code, I'll give it a fix then post the result here.

**Comment by GenghisHan-op** [comment]
I fixed the code like below
```
from machine import Pin
import time
from bleradio import BLERadio

Bpin = machine.Pin(16, machine.Pin.IN, machine.Pin.PULL_DOWN)
Fpin = machine.Pin(17, machine.Pin.IN, machine.Pin.PULL_DOWN)
radio = BLERadio(broadcast_channel=5, observe_channels=[4, 18])
data = None
old_data = None
while True:
    if Fpin.value() == 1:
        data = "F"
        if data != old_data:
            radio.broadcast(data)
            print("Forward")
        old_data = data
    if Bpin.value() == 1:
        data = "B"
        if data != old_data:
            radio.broadcast(data)
            print("Backward")
        old_data = data
    if Fpin.value() == 0 and Bpin.value() == 0:
        data = "S"
        if data != old_data:
            radio.broadcast(data)
            print(data,"  ",old_data," ","STOP")
        old_data = data
    time.sleep(0.1)
```
As you can see, I add machine.Pin.PULL_DOWN when register the Pin, this make the result a lot more stable, no more floating value. Next, I use old_data variable, this will help the program only broadcast when the button changed states (Forward, Backward, Stop). With these fixes, my program run perfectly as I needed. If you have any idea to improve, just tell me.

**Reply by laurensvalk** [reply]
Glad it works!
